src/crimson/osd/shard_services.cc

```diff
@@ -771,49 +781,77 @@ seastar::future<> ShardServices::dispatch_context(
+    m->cluster_osdmap_trim_lower_bound = superblock.cluster_osdmap_trim_lower_bound;
+    m->newest_map = superblock.get_newest_map();
+    auto maybe_handle_mapgap = seastar::now();
     if (first < superblock.cluster_osdmap_trim_lower_bound) {
       logger().info("{}: cluster osdmap lower bound: {} "
-                " > first {}, starting with full map",
-                __func__, superblock.cluster_osdmap_trim_lower_bound, first);
+                    " > first {}, starting with full map",
+                    __func__, superblock.cluster_osdmap_trim_lower_bound, first);
       // we don't have the next map the target wants,
       // so start with a full map.
       first = superblock.cluster_osdmap_trim_lower_bound;
+      maybe_handle_mapgap = load_map_bl(first).then(
+      [&first, &map_message_max, &m](auto&& bl) {
+        m->maps[first] = std::move(bl);
+        --map_message_max;
+        ++first;
+      });
     }
-    return load_map_bls(
-      first, superblock.get_newest_map()
-    ).then([this, &conn](auto&& bls) {
-      auto m = crimson::make_message<MOSDMap>(
-	monc.get_fsid(),
-	osdmap->get_encoding_features());
-      m->cluster_osdmap_trim_lower_bound = superblock.cluster_osdmap_trim_lower_bound;
-      m->newest_map = superblock.get_newest_map();
-      m->maps = std::move(bls);
-      return conn.send(std::move(m));
-    });
-  } else {
-    // See OSDService::send_incremental_map
-    // just send latest full map
-    return load_map_bl(osdmap->get_epoch()
-    ).then([this, &conn](auto&& bl) mutable {
-      auto m = crimson::make_message<MOSDMap>(
-	monc.get_fsid(),
-	osdmap->get_encoding_features());
-      m->cluster_osdmap_trim_lower_bound = superblock.cluster_osdmap_trim_lower_bound;
-      m->newest_map = superblock.get_newest_map();
-      m->maps.emplace(osdmap->get_epoch(), std::move(bl));
-      return conn.send(std::move(m));
+    return maybe_handle_mapgap.then([this, first, last, &map_message_max, &conn, &m] {
+      if (first > last) {
+        // first may be later than last in the case of map gap
+        ceph_assert(!m->maps.empty());
+        return conn.send(std::move(m));
+      }
+      return load_map_bls(
+        first,
+        ((first + map_message_max) > last) ? (first + map_message_max) : last
+      ).then([&conn, &m](auto&& bls) {
+        ssize_t map_message_max_bytes = crimson::common::local_conf()->osd_map_message_max_bytes;
+        for (auto const& [e, val] : bls) {
+          map_message_max_bytes -= val.second.length();
+          if (map_message_max_bytes < 0) {
+            break;
+          }
+          if (val.first == OSDMapService::osdmap_t::FULLMAP) {
+            m->maps.emplace(e, std::move(val.second));
+          } else if (val.first == OSDMapService::osdmap_t::INCMAP) {
+            m->incremental_maps.emplace(e, std::move(val.second));
+          } else {
+            ceph_abort();
+          }
+        }
+        return conn.send(std::move(m));
```

Comment:
If you want to split building from sending, this should probably return the map for send_incremental_map below to send.